src/adapters/uniswap-v3/products/pool/uniswapV3PoolAdapter.ts

```diff
@@ -0,0 +1,354 @@
+import { Erc20Metadata } from '../../../../types/erc20Metadata'
+import { IProtocolAdapter } from '../../../../types/IProtocolAdapter'
+import { Protocol } from '../../../protocols'
+import { PositionManager__factory } from '../../contracts'
+
+const deadline = Math.floor(Date.now() - 1000) + 60 * 10
+
+const positionManagerCommonAddress =
+  '0xC36442b4a4522E871399CD717aBDD847Ab11FE88'
+
+const contractAddresses: Partial<Record<Chain, { positionManager: string }>> = {
+  [Chain.Ethereum]: {
+    positionManager: positionManagerCommonAddress,
+  },
+  [Chain.Arbitrum]: {
+    positionManager: positionManagerCommonAddress,
+  },
+  [Chain.Optimism]: {
+    positionManager: positionManagerCommonAddress,
+  },
+  [Chain.Polygon]: {
+    positionManager: positionManagerCommonAddress,
+  },
+  [Chain.Bsc]: {
+    positionManager: '0x7b8A01B39D58278b5DE7e48c8449c9f4F5170613',
+  },
+  [Chain.Base]: {
+    positionManager: '0x03a520b32C04BF3bEEf7BEb72E919cf822Ed34f1',
+  },
+}
+
+const maxUint128 = BigInt(2) ** BigInt(128) - BigInt(1)
+
+export class UniswapV3PoolAdapter
+  implements IProtocolAdapter, IMetadataBuilder
+{
+  product = 'pool'
+  protocolId: Protocol
+  chainId: Chain
+
+  private provider: ethers.JsonRpcProvider
+
+  constructor({ provider, chainId, protocolId }: ProtocolAdapterParams) {
+    this.provider = provider
+    this.chainId = chainId
+    this.protocolId = protocolId
+  }
+
+  getProtocolDetails(): ProtocolDetails {
+    return {
+      protocolId: this.protocolId,
+      name: 'UniswapV3',
+      description: 'UniswapV3 defi adapter',
+      siteUrl: 'https://uniswap.org/',
+      iconUrl:
+        'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984/logo.png',
+      positionType: PositionType.Supply,
+      chainId: this.chainId,
+    }
+  }
+
+  /**
+   * Update me.
+   * Add logic to build protocol token metadata
+   * For context see dashboard example ./dashboard.png
+   * We need protocol token names, decimals, and also linked underlying tokens
+   */
+  @CacheToFile({ fileKey: 'protocol-token' })
+  async buildMetadata() {
+    throw new Error('Implement me')
+
+    return {}
+  }
+
+  /**
+   * Update me.
+   * Returning an array of your protocol tokens.
+   */
+  async getProtocolTokens(): Promise<Erc20Metadata[]> {
+    throw new Error('Implement me')
+  }
+
+  /**
+   * Update me.
+   * Add logic to get userAddress positions in your protocol
+   */
+  async getPositions({
+    userAddress,
+    blockNumber,
+  }: GetPositionsInput): Promise<ProtocolPosition[]> {
+    const positionsManagerContract = PositionManager__factory.connect(
+      contractAddresses[this.chainId]!.positionManager,
+      this.provider,
+    )
+
+    const balanceOf = await positionsManagerContract.balanceOf(userAddress, {
+      blockTag: blockNumber,
+    })
+
+    const tokenIds = await Promise.all(
+      [...Array(Number(balanceOf)).keys()].map((index) =>
```

Comment:
This could result in a pretty big loop. We might want to come back to this eventually.

Reply:
I'm unsure what the best policy would be here.

I'm pretty sure we could have a single async map that returns `nonZeroPositions` directly without looping three times with something like this.

```
const nonZeroPositions = await Promise.all(
      filterMap([...Array(Number(balanceOf)).keys()], async (index) => {
        const tokenId = await positionsManagerContract.tokenOfOwnerByIndex(
          userAddress,
          index,
          {
            blockTag: blockNumber,
          },
        )

        const position = await positionsManagerContract.positions(tokenId, {
          blockTag: blockNumber,
        })

        if (position.liquidity == 0n) {
          return undefined
        }

        return {
          liquidity: position.liquidity,
          token0: position.token0,
          token1: position.token1,
          fee: position.fee,
          tokenId: tokenIds[index],
        }
      }),
    )
```

I'm unsure whether it'll save any time at all. We would need to run profiling with users with a lot of positions.

Reply:
ive implemented